### ISSUE TITLE
RFC: Explicit setting of engine global additional fields

### ISSUE THREAD

**Issue by NathanKell** [KSP-RO/RealismOverhaul#1669]
I've noticed that many engine configs now explicitly set the global flags when default, e.g.
    %engineTypeMult = 1
    %ignoreMass = False
    %massOffset = 0

I agree this explicitness makes things clearer, but it also increases patch runtime.* Is the benefit worth the cost? Note _I am not taking a position_, this is a genuine question! :)

*This is because if they're not set in the config, then MM tests the config for the flag and, not finding it, moves on. But if it does find it, it then has to apply the full patch, which involves math, variable lookups, and possibly doing that for every CONFIG in the MEC. Considerably slower.

**Comment by jwvanderbeck** [comment]
It's already slow.  A bit slower doesn't seem like a big deal to me.  Percentage wise we are probably talking single digits, and in should all be cached anyway.

**Comment by PhineasFreak** [comment]
Well...that would be me. I like to be explicit for these fields. As @jwvanderbeck says, since only actual parts (that are also engines) are patched, we do not expect a large amount of MM passes (say ~100 if you have a ton of mods that add engines).

Some the RO global patches (that patch max temps, G forces and pressure levels) far outweight the above, as they get applied for all parts.

The worst offender of MM patches is VSR: it applies itself over and over for a lot of part categories (Aero, Command. Electrical, Engines, FuelTanks, Structural and Utility - 7 categories) and for every single part that it finds. If you have 100 parts then VSR will run approximately 100 x 7 = 700 passes (usually it is much worse, with over 7000 to 9000, since i am only accounting the files and to the actual MM patches inside these global part patchers).

**Comment by NathanKell** [comment]
Yep, and we have to pay the overhead cost of checking all root nodes anyway just to _have_ the function.

Max temp, G, and pressure levels don't have to do math at least AIUI, just string comps and sets.

I should look at VSR when I get a chance and see if I can cut that down. Do you know exactly where it is in the repo that's doing that? (The model patches I wrote do a lot of string replacements over all models in all nodes, but I'm not sure of a way to get around that because we do need to check for any instance...)

**Comment by PhineasFreak** [comment]
If you mean the MM patches then they reside under the **PathPatches** folder: https://github.com/VenVen/Stock-Revamp/tree/master/GameData/VenStockRevamp/PathPatches

**Comment by NathanKell** [comment]
Ok so yeah my fault :D

Yeah, we can combine them into a single @PART and single @MODEL,*

**Comment by PhineasFreak** [comment]
Didn't **Ven** originally created these patches? From the file history i can see that you only set up a default MM pass.

**Comment by NathanKell** [comment]
I created them. The repo didn't exist at the time I created them, I just put 'em on the forum. Ven created the repo later.

**Comment by NathanKell** [comment]
(I wrote an Excel sheet to generate the patches.)

**Comment by AnticlockwisePropeller** [comment]
In a similar, but distinct, vein:
I've noticed many engine configs explicitly define all the parameters for that engine before giving an %engineType, whereas other configs simply give an engineType and leave it at that. If the moduleEngineConfigs in the engineType will overwrite whatever was there before, and simply giving an engineType and the few other required parameters (scale, etc.) will work just fine - what's the point in defining all the parameters the engineType will overwrite anyway? Is it just a hangover from before engineTypes existed?

**Comment by pap1723** [comment]
This is a good question that I have also wondered...

**Comment by SirKeplan** [comment]
@AnticlockwisePropeller Yeah it's mostly a hangover. as far as i can tell it makes no functional difference ingame as ModuleEngineConfigs values supersede anything in ModuleEngines/ModuleEnginesRF.

though it's possible it could make a difference to some mods perhaps, as it can leave liquidfuel and oxidiser propellants etc set for the engine, even though this is ignored.

**Comment by PhineasFreak** [comment]
For a start, it does not apply to engine modules that are not targeted by a ModuleEngineConfig (some generic solids, vernier thrusters and other obscure parts).

**Comment by SirKeplan** [comment]
And regarding explicit setting of fields, i don't really mind what we have in the files, though personally i prefer a minimalistic approach, declaring only the fields necessary. this makes sense to me, especially as we might add more optional global fields later there is no need to have them declared but empty everywhere.

**Comment by DRVeyl** [comment]
Problem with explicit setting is when they mis-match the MEC.  The MEC overrides, of course, but this at best only creates confusion.  Current practice [at least on everything I have merged in the past year+] is that for modules covered by the MEC, fields in the ModuleEngine should *not* be present in the RO patch.  This may result in obscure things making it into the MM cache for the ModuleEngine PM, but "don't look at the MM cache's ME PM data, it is meaningless in the presence of the MEC" is an easy diagnostic rule to abide.

Explicit setting of the global flags to their default values [the original Issue] is less problematic; there's far bigger costs to the MM cache build time [TF computations are the single biggest cost by far].  I'm also a fan of not repeating defaults, but when they're parameters to some utilities, it does add clarity.

We can set an internal policy for this if it matters, but closing the issue as there's no new information for >4 years.
